File: 20242Q/poo/taller/taller03/ej1/src/RechargeableSubwayCard.java
public class RechargeableSubwayCard extends SubwayCard {

    private double balance;

    RechargeableSubwayCard(SubwayCentral sc) {
        super(sc);
    }

    public void recharge(double amount) {
        balance += amount;
    }

    protected double rideCost() {
        return getSubwayCentral().getRideCost();
    }

    @Override
    protected boolean canRide() {
        if (balance > 0) {
            return true;
        } else {
            return false;
        }
    }

    @Override
    public void ride() {
        if (canRide() == true) {
            balance -= getSubwayCentral().getRideCost();

            return;
        } else {
            System.err.println("Sin viajes!");

            return;
        }
    }
}

File: 20242Q/poo/laboratorio/tp08/ej6/rechargeable_subway_card.rb
require_relative 'subway_card'

class RechargeableSubwayCard < SubwayCard
  def initialize(subway_central)
    @subway_central = subway_central
    @balance = 0
  end
  
  def recharge(amount)
    raise 'Invalid amount' unless amount.is_a?(Numeric) && amount.positive?
    @balance += amount
  end

  private def ride_cost
    @subway_central.ride_cost
  end

  private def ride?
    (@balance - ride_cost).positive?
  end

  def ride
    raise 'Not enough funds' unless ride?
    @balance -= ride_cost
  end

  def to_s
    "Rechargable #{super} [Balance: $#{@balance}]"
  end
end
